updater.go

```diff
@@ -244,19 +285,105 @@ func (u *Updater) AppendHeaderAt(fh *FileHeader, offset int64) (io.Writer, error
 		return nil, err
 	}
 
-	if offset < 0 {
-		offset = u.dirOffset
-	}
 	// Offset should match existing header offsets or equal to directory offset.
 	var offsetExists bool
-	for i, h := range u.dir {
-		if h.offset == uint64(offset) {
-			offsetExists = true
-			// Delete the corresponding header.
-			u.dir = append(u.dir[:i], u.dir[i+1:]...)
-			break
+
+	existingFileIndex := -1
+	if offset < 0 {
+		// Append the file to the zip or replace the existing one
+		for i, h := range u.dir {
+			if h.FileHeader.Name == fh.Name {
+				offset = int64(h.offset)
+				existingFileIndex = i
+				break
+			}
+		}
+
+		// File does not exist inside the zip, so we append it
+		if existingFileIndex < 0 {
+			offset = u.dirOffset
+		}
+	} else {
+		// This is for manually writing data into a certain offset
+		for i, h := range u.dir {
+			if h.offset == uint64(offset) {
+				offsetExists = true
+				// Delete the corresponding header.
+				u.dir = append(u.dir[:i], u.dir[i+1:]...)
+				break
+			}
 		}
 	}
+
+	if existingFileIndex >= 0 {
+		// If the file exists we will remove it and reinsert it at the end of the file.
+		// The existing files data will be relocated first to use the leftover space.
+
+		// This diagram shows the internal structure of a zip file, whose i-th file
+		// we want to overwrite. The annotations represent some of the variables that
+		// are used to implement this feature.
+		//
+		//                       deletedDataSize = nextOffset - existingFileOffset
+		//                      ┌────────────────┐
+		//                      ▼                ▼
+		// ┌───┬────────────┬───┬───┬────────────┬───┬────────────┬───┬───────┐
+		// │   │            │   │   │            │   │            │   │░░░░░░░│
+		// │h_0│   File 0   │...│h_i│   File i   │h_ │  File i+1  │...│░░Dir░░│
+		// │   │    data    │   │   │    data    │i+1│    data    │   │░░░░░░░│
+		// │   │            │   │   │            │   │            │   │░░░░░░░│
+		// └───┴────────────┴───┴───┴────────────┴───┴────────────┴───┴───────┘
+		//                      ▲                ▲                    ▲
+		//                      │                │                    │
+		//                   existingFileOffset  nextFileOffset       Dir Offset
+
+		existingFileOffset := u.dir[existingFileIndex].offset
+		var deletedDataSize uint64
+
+		if existingFileIndex+1 < len(u.dir) {
+			nextFileOffset := u.dir[existingFileIndex+1].offset
+
+			deletedDataSize = nextFileOffset - existingFileOffset
+			filesInZip := len(u.dir)
+
+			// Rewind the existing data by the deleted data size
+			err := u.rewindData(int64(existingFileOffset), int64(deletedDataSize))
+			if err != nil {
+				return nil, err
+			}
+
+			for i := existingFileIndex + 1; i < filesInZip; i++ {
+				// Update the file offsets in their headers, to match their new positions
+				u.dir[i].offset = u.dir[i].offset - uint64(deletedDataSize)
```

Comment:
Also need to delete the extra data of the zip64 file header, or the header offset of zip64 file won't be updated correctly.
The zip64 header extra data is generated at [here](https://github.com/Meteologica/zip/blob/cd84ce91987908277e6ab8a5f2109f762b01bff4/writer.go#L102C3-L117C40).

```go
for i := existingFileIndex + 1; i < filesInZip; i++ {
	// Update the file offsets in their headers, to match their new positions
	u.dir[i].offset = u.dir[i].offset - uint64(deletedDataSize)
	// Delete the existing zip64 header extra data
	u.dir[i].Extra = nil
}
```